scipy/integrate/_quadrature.py

```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        \left\{3-\frac{x_2-x_1}{x_3-x_1}\right\} y_1 + \
+        \left\{3 + \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} + \
+        \frac{x_2-x_1}{x_3-x_1}\right\} y_2 - \
+        \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} y_3\right]
+
+    The integral between :math:`x_2` and :math:`x_3` is given by swapping
+    appearances of :math:`x_1` and :math:`x_3`. The integral is estimated
+    separately for each subinterval and then cumulatively summed to obtain
+    the final result.
+    
+    For samples that are equally spaced, the result is exact if the function
+    is a polynomial of order 3 or less [1]_ and the number of subintervals
+    is even. Otherwise, the integral is exact for polynomials of  order 2 or
+    less. 
+
+    References
+    ----------
+    .. [1] Wikipedia page: https://en.wikipedia.org/wiki/Simpson's_rule
+    .. [2] Cartwright, Kenneth V. Simpson's Rule Cumulative Integration with
+            MS Excel and Irregularly-spaced Data. Journal of Mathematical
+            Sciences and Mathematics Education. 12 (2): 1-9
+
+    Examples
+    --------
+    >>> from scipy import integrate
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> x = np.linspace(-2, 2, num=20)
+    >>> y = x**2
+    >>> y_int = integrate.cumulative_simpson(y, x=x, initial=0)
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(x, y_int, 'ro', x, x**3/3 - (x[0])**3/3, 'b-')
+    >>> ax.grid()
+    >>> plt.show()
+
+    The output of `cumulative_simpson` is similar to that of iteratively
+    calling `simpson` with successively higher upper limits of integration, but
+    not identical.
+
+    >>> def cumulative_simpson_reference(y, x):
+    ...     return np.asarray([integrate.simpson(y[:i], x=x[:i])
+    ...                        for i in range(2, len(y) + 1)])
+    >>>
+    >>> rng = np.random.default_rng(354673834679465)
+    >>> x, y = rng.random(size=(2, 10))
+    >>> x.sort()
+    >>>
+    >>> res = integrate.cumulative_simpson(y, x=x)
+    >>> ref = cumulative_simpson_reference(y, x)
+    >>> equal = np.abs(res - ref) < 1e-15
+    >>> equal  # not equal when `simpson` has even number of subintervals
+    array([False,  True, False,  True, False,  True, False,  True,  True])
+
+    This is expected: because `cumulative_simpson` has access to more
+    information than `simpson`, it can typically produce more accurate
+    estimates of the underlying integral over subintervals.
+
+    """
+    y = _ensure_float_array(y)
+
+    # validate `axis` and standardize to work along the last axis
+    original_shape = y.shape
+    try:
+        y = np.swapaxes(y, axis, -1)
+    except IndexError as e:
+        message = f"`axis={axis}` is not valid for `y` with `y.ndim={y.ndim}`."
+        raise ValueError(message) from e
+
+    if y.shape[-1] < 3:
+        raise ValueError(
+            "At least 3 points are required along the axis of integration "
+            "to use the composite Simpson's method."
+        )
```

Comment:
This is not an error in `simpson` 
```python3
integrate.simpson([1], x=[1])
# 0.0
```

or the analogous case in `cumulative_trapezoid`:
```python3
from scipy import integrate
integrate.cumulative_trapezoid([1], x=[1], initial=0)
# array([0.])
```

So I think we should make this work when `y.shape[-1] > 0`, and raise the error here when `y.shape[-1] = 0`. Because of the function's behavior of reducing the length by one, I don't think there is a correct result to be returned when `y.shape[-1] = 0` and `initial` is not specified. With `initial` specified, it could be argued that an empty array would be the correct result, but I think this would be confusing. (If only these could follow normal broadcasting rules and always return an output with the same shape as `y` : / It would all be so much simpler!)

Reply:
To address this, shall we do the following then:
0< y.shape[-1] < 3 ---> return result from cumulative_trapezoid
y.shape[-1] --> raise error

This would match the behaviour from simpson. Simpson also just returns trapezoid result when y.shape[-1] = 2

Reply:
Yeah, that's a good idea. Actually, just let the function this calls raise the error for 0 length (and make sure it's understandable in this context, too).

Reply:
Addressed in #19709
cumulative_trapezoid does not have an explicit error for when y.shape[-1] <=0. Hence added one in in cumulative_simpson: https://github.com/scipy/scipy/pull/19709/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR1042

Reply:
Yeah, so then why not add the error to `cumulative_trapezoid`, since it's desirable there, too?